components/datetime/src/pattern/parser.rs

```diff
@@ -0,0 +1,118 @@
+use super::error::Error;
+use super::{Pattern, PatternItem};
+use crate::fields::{Field, FieldLength, FieldSymbol};
+use std::convert::TryFrom;
+
+enum State {
+    Symbol {
+        symbol: FieldSymbol,
+        byte: u8,
+        start_idx: usize,
+    },
+    Literal {
+        start_idx: usize,
+    },
+}
+
+pub struct Parser<'p> {
+    source: &'p [u8],
+    state: State,
+    items: Vec<PatternItem>,
+    ptr: usize,
+}
+
+impl<'p> Parser<'p> {
+    pub fn new(source: &'p [u8]) -> Self {
+        Self {
+            source,
+            state: State::Literal { start_idx: 0 },
+            items: Vec::with_capacity(source.len()),
```

Comment:
Possibly more items than you need; ok?

Reply:
I think so. The difference shouldn't be that large (usually we have zero or one literals and in most cases they are very short strings).

Reserving that capacity is an easy approximation protecting us from having to reallocate as the vec grows. I can try removing that and see the perf impact.